core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -224,11 +224,19 @@ class FileMessageSet private[kafka](@volatile var file: File,
       }
     }
 
-    // We use the offset seq to assign offsets so the offset of the messages does not change.
-    new ByteBufferMessageSet(
-      compressionCodec = this.headOption.map(_.message.compressionCodec).getOrElse(NoCompressionCodec),
-      offsetSeq = offsets,
-      newMessages: _*)
+    if (_size.get() > 0 && newMessages.size == 0) {
+      // This indicates that the message is too large. We just return all the bytes in the file message set.
+      val buffer = ByteBuffer.allocate(_size.get())
+      channel.read(buffer, start)
+      buffer.rewind()
+      new ByteBufferMessageSet(buffer)
```

Comment:
Would it be better to simply return the FileMessageSet itself in this case? We will need to change toMessageFormat() to return MessageSet instead.